### ISSUE TITLE
NAT Model, Act I

### ISSUE THREAD

**Issue by lukasjuhrich** [agdsn/pycroft#330]
Child of #73.

# Goal
The schema's hard relations (_not_ including triggers!) should be implemented as intended in shreyders POC: ![NAT Schema](https://agdsn.me/~lukasjuhrich/nat_schema.png). Tests are required. Target branch is `nat`.

# Tasks
- [ ] Add all dummy tables
- [ ] Add a primitive Test for all the tables and a suitable test fixture
- [ ] Add the columns w/ correct FKey relations and other constraints
- [ ] Enrich the test fixture with data and add a „dummy“ test

**Comment by lukasjuhrich** [comment]
We have to cover the following use cases:

- [x] Legacy/Wired: Access-Switch is configured to ask Hades, Hades responds with 666 or the /Dorm-VLAN/.
- [ ] NAT/Wired: Access-Switch is configured to ask Hades, Hades responds with 666 or the /User-VLAN/ (corresponding to his ISID)
- [ ] NAT/Wireless: WLAN-Controller gets 802.1X auth (⇒`User.login`), needs corresponding ISID(?)

**Comment by sebschrader** [comment]
Our current and future switches, access points, and WLAN controllers don't support dynamic I-SID assignments via RADIUS. Only VLANs can be assigned dynamically via RADIUS. VLANs can be associated statically with I-SIDs on the switches however. So users can be indirectly assigned dynamically to an I-SID via RADIUS through the use of VLANs.

The VLANs have only local significance on the particular switch. So the VLAN ID to use depends on the NAS (Network Access Server) in RADIUS parlance, the user tries to connect to. We need at least a wired VLAN ID on their home switch and wireless VLAN ID on the WLAN controller for each user.

In summary:
* Each user has a layer 2 domain
* The L2 Domain has an I-SID associated with it
* The L2 Domain has multiple NAS-specific VLAN IDs

**Comment by lukasjuhrich** [comment]
I would argue that not every user „has“ a layer 2 domain – instead, layer 2 domains (`internal_network`) can be associated to a user. Remember that only a subset of our users will join the new network model, others may still be oblivious to that information.

## Proposition

Okay, so I propose the following changes:

- [ ] Add a `isid(int isid, string? name)` relation representing the pool of all available isids
- [ ] Add an FKey `internal_network.isid → isid.isid`
- [ ] Add a translation `isid_vlan(int isid, int nas_id, something vlan)` (`isid, nas` primary, `vlan` unique)

## Left to decide
- [ ] Do we want a constraint that every isid must have a complete set of mappings with every nas?
- [ ] Do we want a hard FKey `isid_vlan.vlan_id → vlan.id`?
- [ ] If yes, do we want a constraint checking that each vlan is exclusively associated to either a subnet (legacy-model) or an isid (nat-model)?

I'd argue that the third point is overkill, but the second might be useful.  By having ~nullable~ possibly empty backrefs `VLAN.subnets`, `VLAN.isids`, the UI can easily inform about which „kind“ of vlan this is (see #334).

**Comment by lukasjuhrich** [comment]
Since I lost the overview about which checks etc. to implement, I reconstructed the feature list of the model in [this `.org`-gist](https://gist.github.com/lukasjuhrich/77161814c2da368a227713d00213c94b).
Also interesting for #331.

I'll try to split those features up to get a precise definition of what has to be done for an initial deployment of the (unfinished) model.
